track participant changes and send participant change email

jmartin-sul opened the pull request:
## Why was this change made?

close #596 
close #781 

## How was this change tested?

unit tests, manual testing on local instance (see screenshots)

<img width="1113" alt="Screen Shot 2021-01-14 at 8 45 46 PM" src="https://user-images.githubusercontent.com/7741604/104682604-f4b69400-56a9-11eb-9e24-9203dd20580e.png">
<img width="1187" alt="Screen Shot 2021-01-14 at 8 46 17 PM" src="https://user-images.githubusercontent.com/7741604/104682607-f6805780-56a9-11eb-8b97-1840c02a10c1.png">


## Which documentation and/or configurations were updated?

n/a

jmartin-sul commented:
will likely have to rebase if #846 and/or #849 get merged first

jcoyne commented on spec/requests/edit_collection_spec.rb:
```diff
@@ -108,6 +114,57 @@
+            expect(event_description).to match(/Removed depositors: #{removed_depositor.sunetid}/)
+          end
+
+          # rubocop:disable RSpec/NestedGroups
```
This seems like a sane number of nested groups, you could bump up this threshold in .rubocop.yml

jmartin-sul commented on app/components/collections/history_component.html.erb:
```diff
@@ -13,7 +13,11 @@
         <td><%= I18n.l(event.created_at, format: :abbr_month) %></td>
-        <td><%= event.description %></td>
+        <td>
+          <% (event.description || '').split("\n").each do |desc_line| %>
```
will do, wasn't aware of that method, thanks...

jmartin-sul commented on app/models/collection_change_set.rb:
```diff
@@ -20,22 +24,51 @@ def added_reviewers
+  end
+
+  def participant_changes
+    {}.tap do |h|
```
ah, that's sorta cruft from earlier in my implementation when the caller was asking for `participant_changes` and doing the formatting instead of calling `particpant_change_description`.  will get rid of the extra indirection...

jmartin-sul commented:
OK, I think I've incorporated all of the suggested changes.  I also rebased this, _but_ I don't think I fully took the spirit of the refactoring on which I rebased (from #846).  In particular, I kept much of the new spec code I wrote in the same place in `edit_collection_spec.rb`, instead of moving it to `collection_observer_spec.rb`.  I'm happy to revisit this and to continue refactoring that code, either in this PR or in a follow-on.  However, I'd planned to take a half-day today to run some errands, and I need to take off to do that.  I'm happy to circle back on this later this evening, or on Tuesday.

jmartin-sul commented:
I also didn't scrutinize the post-rebase diff quite as much as I normally would -- if this still isn't merged if/when I get back to working on it, I'll give things a more thorough once-over.

jcoyne commented on app/models/collection.rb:
```diff
@@ -55,6 +57,7 @@ def purl
     end
 
     after_transition to: :version_draft, do: CollectionObserver.method(:after_update_published)
+    after_transition on: :update_metadata, do: CollectionObserver.method(:on_update_metadata)
```
This is unnecessary, there already is a hook that you should be using called after_update_published

jmartin-sul replied:
fixed (wasn't sure on friday whether i should be using that same hook, or whether i should've caught all metadata updates)

jmartin-sul commented:
ok, incorporated the most recent bit of feedback, and i think i've followed through on the refactoring started in #846.  i left the tests around updating event description in `edit_collection_spec.rb`, since that still happens in the `before_transition` hook on `collection.rb`.  also rebased again on `main` (as of a few minutes ago).

mjgiarlo reviewed (approved):
LGTM
